### ISSUE TITLE
`semi` rule reports incorrect line for missing semicolon

### ISSUE THREAD

**Issue by iancmyers** [eslint/eslint#451]
Consider the following example:

``` js
doAsyncThing(function (result) { // This line is reported as having a missing semicolon.
    doStuffWithResult(result);
}) // This is the line with the missing semicolon
```

I think this is happening because the `ExpressionStatement` is the node being reported and that node starts on the first line.

I brought up on a different issue the idea of being able to report tokens. It seems like we either need to be able to report tokens _or_ on a rule-by-rule basis configure whether the line reported is the node start line or the node end line. I'm not sure how else we can get accurate line numbers in cases like this otherwise.
